src/rust_connection/fd_read_write.rs

```diff
@@ -175,82 +185,48 @@ impl<W: WriteFD> WriteFD for BufWriteFD<W> {
             }
         }
 
-        if buf.len() >= self.data_buf.capacity() {
-            // At this point the buffer is empty
-            match self.inner.write(buf, &mut self.fd_buf) {
-                Ok(n) => Ok(n),
-                Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
-                    let n_to_write = buf
-                        .len()
-                        .min(self.data_buf.capacity() - self.data_buf.len());
-                    let _ = self.data_buf.write(&buf[..n_to_write]).unwrap();
-                    // Return `Ok` because some data has been buffered,
-                    // so from the outside it is seen as a successful write.
-                    Ok(n_to_write)
-                }
-                Err(e) => Err(e),
-            }
+        if to_write_length >= self.data_buf.capacity() {
+            // Write is larger than the buffer capacity, thus we just flushed the buffer. This
+            // means that at this point the buffer is empty. Write directly to self.inner. No data
+            // is copied into the buffer, since that would just mean that the large write gets
+            // split into multiple smaller ones.
+            assert!(self.data_buf.is_empty());
+            write_inner(&mut self.inner, &mut self.fd_buf)
```

Comment:
We still need to copy the data into the buffer if `write_inner` returns `WouldBlock`. Otherwise, `write`/`write_vectored` could return `WouldBlock` after `poll` said that the stream is writable.

Reply:
Hm... I was going to propose the following instead
```diff
@@ -260,12 +261,7 @@ impl<W: WriteFD> WriteFD for BufWriteFD<W> {
 
 impl<T: WriteFD> Poll for BufWriteFD<T> {
     fn poll(&mut self, read: bool, write: bool) -> Result<(bool, bool)> {
-        if write && self.data_buf.len() != self.data_buf.capacity() {
-            // Avoid blocking poll if write buffer is not full.
-            Ok((false, true))
-        } else {
-            self.inner.poll(read, write)
-        }
+        self.inner.poll(read, write)
     }
 }
 
```
But that's not really enough, since `flush_buffer` could still make partial progress.

How about explicitly allowing spurious wakeups for `poll()`? I.e., even though `poll()` returns writability, the writer may still return `WouldBlock`. That would just mean that we document my patch as "this is okay". ;-)

Reply:
`man poll` says

> BUGS
>       See the discussion of spurious readiness notifications under the BUGS section of select(2).

`man select` says:

>       Under  Linux,  select() may report a socket file descriptor as "ready for reading", while nevertheless a subsequent
>       read blocks.  This could for example happen when data has arrived but upon examination has wrong  checksum  and  is
>       discarded.   There  may be other circumstances in which a file descriptor is spuriously reported as ready.  Thus it
>       may be safer to use O_NONBLOCK on sockets that should not block.

Reply:
In other words, Linux is not sane :wink:

Then I think we can merge this PR as is and I'll and open a new one to fix the remaining code.

Reply:
Oh, right, we speculated about this in the other PR. I forgot.